proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,208 @@
+# Unlock Existentials for All Protocols
+
+* Proposal: [SE-NNNN](NNNN-unlock-existential-types-for-all-protocols.md)
+* Authors: [Anthony Latsis](https://github.com/AnthonyLatsis), [Filip Sakel](https://github.com/filip-sakel), [Suyash Srijan](https://github.com/theblixguy)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#33767](https://github.com/apple/swift/pull/33767)
+
+
+### Introduction
+
+Swift currently offers the ability for protocols that meet certain criteria to be used as types. Trying to use an unsupported protocol as a type yields the error `[the protocol] can only be used as a generic constraint because it has 'Self' or associated type requirements`. This proposal aims to relax this artificial constraint imposed on such protocols.
+
+
+### Motivation
+
+Currently, any protocol that has [non-covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) `Self` references or associated type requirements is not allowed to be used as a type. Initially, this restriction reflected technical limitations (as Joe states [here](https://forums.swift.org/t/lifting-the-self-or-associated-type-constraint-on-existentials/18025)); however, such limitations have now been alleviated. As a result, users are left unable to utilize a powerful feature for certain protocols. That’s evident in a plethora of projects. For instance, the Standard Library has existential types such as [`AnyHashable`](https://developer.apple.com/documentation/swift/anyhashable) and [`AnyCollection`](https://developer.apple.com/documentation/swift/anycollection) and SwiftUI has [`AnyView`](https://developer.apple.com/documentation/swiftui/anyview). 
+
+Generics are most often the best mechanism for type-level abstraction, which relies on the compiler knowing type information during compilation. However, type information is not always a gurantee, which is why the value-level abstraction of existential types is extremely useful in some cases.
+
+One such case is heterogenous collections, which require value-level abstraction to store their elements of various types:
+
+```swift
+protocol Identifiable {
+    associatedtype ID: Hashable 
+    
+    var id: ID { get }
+}
+
+struct AnyIdentifiable { 
+    typealias ID = AnyHashable
+    
+    var id: ID { ... }
+}
+
+let myIdentifiables: [AnyIdentifiable]
```

Comment:
I think we also need to talk about known associated types as well in this section.

Reply:
Right! Especially if we consider that this is the problem the original proposal addressed.

Reply:
Okey, I've added it!

---

Morover, the availability of an existential types for a given protocol is sometimes quite unintuitive. That is, today, a protocol qualifies for an existential type provided that it lacks any `Self` or associated type requirements; however, the associated types of a protocol can be fixed via the same-type constraint. As a result, [post](https://forums.swift.org/t/making-a-protocols-associated-type-concrete-via-inheritance/6557) after [post](https://forums.swift.org/t/constrained-associated-type-on-protocol/38770) has been created asking for this restriction's removal:

 ```swift
 protocol User: Identifiable
   where ID == UUID {
   
   var username: Strng { get }
 }
   
 let myUsers: [User] ❌
 // This is forbidden today
 // for no apparent reason.
 ```